File: app/src/main/kotlin/app/guitar/app/RhythmUnitState.kt
package app.guitar.app

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import app.guitar.audio.AudioEngine
import app.guitar.theory.RhythmUnit
import app.guitar.theory.RhythmUnits
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * Playback for the Rhythmic Units screen: loops one selected one-beat unit,
 * clicking a synthesized woodblock-like tick on each note onset (the first onset
 * of every beat accented). BPM-controllable. Reuses the drum machine's
 * mixer-clock lookahead scheduling ([AudioEngine.playSamplesAt]) so timing is
 * sample-accurate and never drifts.
 */
class RhythmUnitState(
    private val audio: AudioEngine,
    private val scope: CoroutineScope,
) {
    var selectedId by mutableStateOf<String?>(null)
        private set
    var isPlaying by mutableStateOf(false)
        private set
    var bpm by mutableStateOf(30)
        private set

    private var job: Job? = null

    // Pre-rendered click buffers: a short exponentially-decaying tone. Accent (beat 1)
    // is a touch higher + louder so the downbeat is felt.
    private val click = synthClick(2000.0, 45)
    private val accentClick = synthClick(2800.0, 45)

    val selected: RhythmUnit? get() = selectedId?.let { RhythmUnits.byId(it) }

    /** Tap a unit card: switch to it and (re)start its loop; tapping the one that's
     *  already playing stops. */
    fun select(id: String) {
        if (selectedId == id && isPlaying) { stop(); return }
        selectedId = id
        restart()
    }

    fun toggle() {
        if (isPlaying) stop() else if (selectedId != null) start()
    }

    // Named changeBpm (not setBpm) to avoid clashing with the generated `bpm` setter.
    fun changeBpm(v: Int) { bpm = v.coerceIn(10, 300) }   // loop re-reads bpm each beat

    // Clear isPlaying before restarting so start()'s `if (isPlaying) return` guard
    // doesn't swallow an instant switch to another unit while one is playing.
    private fun restart() { cancelJob(); isPlaying = false; start() }

    fun start() {
        val u = selected ?: return
        if (isPlaying) return
        isPlaying = true
        job = scope.launch { loop(u) }
    }

    fun stop() {
        cancelJob()
        isPlaying = false
    }

    private fun cancelJob() { job?.cancel(); job = null }

    private suspend fun loop(u: RhythmUnit) {
        val sr = audio.sampleRate
        val fractions = u.clickFractions()   // rests produce no click
        var nextBeatNanos = System.nanoTime()
        while (isPlaying) {
            val beatMs = 60_000.0 / bpm
            val baseDelayMs = ((nextBeatNanos - System.nanoTime()) / 1_000_000.0).coerceAtLeast(0.0)
            for (f in fractions) {
                val onsetDelayMs = baseDelayMs + f * beatMs
                val delayFrames = (onsetDelayMs * sr / 1000.0).toInt().coerceAtLeast(0)
                // Accent the downbeat click (a note landing on the beat), not merely the
                // first click — so a unit that rests on beat 1 isn't falsely accented.
                if (f == 0.0) audio.playSamplesAt(accentClick, 1.0f, delayFrames)
                else audio.playSamplesAt(click, 0.72f, delayFrames)
            }
            nextBeatNanos += (beatMs * 1_000_000).toLong()
            delay((baseDelayMs + beatMs).toLong().coerceAtLeast(1))
        }
    }

    private fun synthClick(freqHz: Double, ms: Int, sr: Int = audio.sampleRate): FloatArray {
        val n = sr * ms / 1000
        val buf = FloatArray(n)
        val w = 2.0 * Math.PI * freqHz / sr
        for (i in 0 until n) {
            val env = Math.exp(-6.0 * i / n)   // fast percussive decay
            buf[i] = (Math.sin(w * i) * env * 0.7).toFloat()
        }
        return buf
    }
}

File: chorect-web/src/app/rhythmUnitState.ts
// Playback for the Rhythmic Units screen — mirror of app/.../RhythmUnitState.kt.
// Loops one selected one-beat unit, clicking a synthesized woodblock-like tick on
// each note onset (downbeat accented), at a controllable BPM. Schedules on the
// AudioContext clock (playSamples(…, whenSec)) so timing never drifts; a token
// counter replaces Kotlin's Job cancellation.

import { WebAudioEngine } from "../audio";
import { clickAt } from "./woodClick";
import { RhythmUnit, rhythmUnitById, clickFractions } from "../theory";

const sleep = (ms: number) => new Promise<void>((r) => setTimeout(r, ms));

export interface RhythmUnitDeps {
  audio: WebAudioEngine;
  onChange: () => void;
}

export class RhythmUnitState {
  selectedId: string | null = null;
  isPlaying = false;
  bpm = 30;

  private token = 0;

  constructor(private deps: RhythmUnitDeps) {}

  /** Clicks are built lazily at the ENGINE's rate — see clickAt(). */
  private get click(): Float32Array { return clickAt(2000, 45, this.deps.audio.sampleRate); }
  private get accentClick(): Float32Array { return clickAt(2800, 45, this.deps.audio.sampleRate); }

  get selected(): RhythmUnit | undefined {
    return this.selectedId ? rhythmUnitById(this.selectedId) : undefined;
  }

  /** Tap a card: switch to it and (re)start; tapping the playing one stops.
   *  Switching while another unit plays swaps instantly (no stop-then-play). */
  select(id: string): void {
    if (this.selectedId === id && this.isPlaying) { this.stop(); return; }
    this.selectedId = id;
    this.stopLoop();          // cancel any running loop (bumps the token)
    this.isPlaying = false;   // clear the guard so start() proceeds with the new unit
    this.start();
  }

  toggle(): void {
    if (this.isPlaying) this.stop();
    else if (this.selectedId) this.start();
  }

  setBpm(v: number): void {
    this.bpm = Math.round(Math.min(Math.max(v, 10), 300));   // loop re-reads bpm each beat
    this.deps.onChange();
  }

  start(): void {
    const u = this.selected;
    if (!u || this.isPlaying) return;
    this.isPlaying = true;
    this.deps.onChange();
    void this.loop(u, ++this.token);
  }

  stop(): void {
    this.stopLoop();
    this.isPlaying = false;
    this.deps.onChange();
  }

  private stopLoop(): void { this.token++; }

  private async loop(u: RhythmUnit, token: number): Promise<void> {
    const fractions = clickFractions(u);   // rests produce no click
    let nextBeat = this.deps.audio.now();
    while (this.isPlaying && token === this.token) {
      const beatSec = 60 / this.bpm;
      const base = Math.max(nextBeat - this.deps.audio.now(), 0);
      fractions.forEach((f) => {
        const whenSec = this.deps.audio.now() + base + f * beatSec;
        // Accent the downbeat click (a note on the beat), not merely the first click.
        if (f === 0) this.deps.audio.playSamples(this.accentClick, 1.0, whenSec);
        else this.deps.audio.playSamples(this.click, 0.72, whenSec);
      });
      nextBeat += beatSec;
      await sleep((base + beatSec) * 1000);
    }
  }
}

